tidymodels-regression-part2.qmd

```diff
@@ -0,0 +1,274 @@
+  args_list = arx_args_list(
+    lags = list(death_rate = c(1), case_rate = c(0,1)),
+    ahead = 0L
+  )
+) 
+ca_arx_pred_jan_22$predictions
+```
+
+We could easily add more lags for the case and death rates into the function . All that we got to do is add a couple of choice numbers to the `lags` argument. This is infinitely preferable to the (longwinded) alternative of adding them one by one to `step_mutate()`. So let's go ahead and try adding various lags for each of case and death rate to `arx_forecaster()`:
+
+```{r, warning = FALSE}
+lots_of_lags_ca <- arx_forecaster(
+  ca_subset_plus_row %>% select(-lag_1_death_rate), 
+  outcome = "death_rate", 
+  predictors = c("case_rate", "death_rate"),
+  args_list = arx_args_list(
+    lags = list(case_rate = c(0, 1, 2, 3, 4, 7), death_rate = c(1, 2, 3, 4, 7)),
+    ahead = 0L
+  )
+) 
+lots_of_lags_ca$predictions
+```
+
+The other major benefit of the above forecaster is that it is equipped to handle panel data. That means that for our example we could input several other states worth of data, and get a prediction for each state. For example, we can try plugging in the entirety of the original `case_death_rate_subset` and get a prediction of the death rate on Jan. 1, 2022 for each state. 
+
+```{r, warning = FALSE}
+all_the_states <- arx_forecaster(
+  case_death_rate_subset, 
+  outcome = "death_rate", 
+  predictors = c("case_rate", "death_rate"),
+  args_list = arx_args_list(
+    lags = list(c(0, 1, 2, 3, 4, 7), c(1, 2, 3, 4, 7)),
+    ahead = 1L
+  )
+) 
+all_the_states$predictions
+```
+
+Neat! We'll learn more about this and other forecasters in later chapters.
+
+## Interactive plot of predictions {#sec-interactive-plot}
+
+Let's finish off by producing an interactive `plotly` choropleth map of our above predictions (along with the 90% predictive intervals) to get a sense of how they may be geospatially related.
+
+First we'll ready our data. The key part is the extraction of the quantiles that make up the 90% predictive intervals. These are by default contained inside a distribution (`.pred_distn`) which we must unnest to get to.
+
+```{r}
+# Rename df of predictions to shorten up
+all_the_states_df <- all_the_states$predictions
+
+# Extract nested quantiles
+all_the_states_df <- all_the_states_df %>% 
+  mutate(q = nested_quantiles(.pred_distn)) %>% 
+  unnest(q) %>%
+  pivot_wider(names_from = tau, values_from = q)
+
+```
+
+Then, we simply adapted the "Customize choropleth code" from the [plotly website](https://plotly.com/r/choropleth-maps/) to our data. The major changes are that we specified what we would like to show when we hover (the predictions and the corresponding 90% predictive interval). As well, we modified the trace (a trace is just a layer to add to the plot with its own data and visualization components), so that the predictions determine the colour of the state and the locations are based on the state abbreviations in `geo_value`.
+
+```{r}
```

Comment:
`plotly()` it self is kind of ugly. There's a `ggplot2` version called `ggplotly`. Can you try remaking it with this?

Reply:
After trying to make this work with `ggplotly`, I have decided to leave this as plotly-based because it works without having to install multiple packages, it looks ok, and it doesn’t throw warnings or errors.

Reasons I have found as to why not go for `ggplotly()`…  A common way to produce a map plot in `ggplot2` to use `ggplot(sf_state_data) + geom_sf()` (where the inputted data to produce the map of the US split by states is a simple features data structure). However, I have not been able to find a nice map of the us in sf format. The `maps::map("state”)` plot that people often use here does not include HI and AK. The `albersusa` and a few other dependent packages that are commonly used to query geo-spatial data as sf data format are/have been recently retired… Another package I tried has Hawaii unusually far off from the rest of the states. `urbnmapr` may be one possibility to look more into (I actually like how the state plot looks), but I haven’t seen anyone else use it to generate a dynamic plot with `ggplotly()`… When I try to make this work, I get a warning that the text and frame aesthetics, which previously worked with `geom_sf` are unknown (the text when hovering and the frame of the animation by year). So, I am not sure the current version of `geom_sf` works to produce a dynamic plot with `plotly` & to get this to work would likely require multiple packages & more time than is probably worth spending on this right now.